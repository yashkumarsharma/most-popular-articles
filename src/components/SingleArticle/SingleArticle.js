import PropTypes from 'prop-types'
import Skeleton from 'react-loading-skeleton'

import { getLargeImage, formatDate } from '../../utils'
import Tags from '../Tags'

const SingleArticle = (props) => {
  const {
    article = {},
    isLoading = false,
  } = props

  const {
    title,
    abstract,
    published_date,
    byline,
    media,
    adx_keywords = '',
    des_facet,
    org_facet,
    per_facet,
    geo_facet,
  } = article

  const image = getLargeImage(media)
  const keywords = adx_keywords.split(';').filter(Boolean)

  return (
    <>
      {!!isLoading && (
        <Skeleton
          className='bg-gray-400 p-80 flex h-full w-full'
        />
      )}
      {!isLoading && (
        <div className='overflow-y-auto md:w-3/5 bg-gray-100 p-4 md:mt-0 mt-4' data-testid='article-details'>
          <h3 className='text-2xl font-semibold mb-2'>{title}</h3>
          <div className='mb-4 text-center'>
            <img src={image} alt={title} className='article-image w-full mx-auto rounded-lg shadow-lg' />
            <div className='flex flex-row justify-between max-w-2/3 mx-auto'>
              <div className='text-sm text-gray-600' data-testid='date-details'>
                Published on {formatDate(published_date)}
              </div>
              <div className='text-sm text-gray-600'>{byline}</div>
            </div>
          </div>
          <p className='mb-4'>{abstract}</p>
          <Tags
            title='Descriptive Terms'
            content={des_facet}
          />
          <Tags
            title='Organizations'
            content={org_facet}
          />
          <Tags
            title='Persons'
            content={per_facet}
          />
          <Tags
            title='Locations'
            content={geo_facet}
          />
          <Tags
            title='Keywords'
            content={keywords}
          />
        </div>
      )}
    </>
  )
}

SingleArticle.propTypes = {
  article: PropTypes.shape({}),
  isLoading: PropTypes.bool,
}

export default SingleArticle
