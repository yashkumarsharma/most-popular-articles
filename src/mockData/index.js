// eslint-disable-next-line import/prefer-default-export
export const mockArticles = [{
  'uri': 'nyt://article/7d22b056-2e3b-5778-b1c4-526f651778b9',
  'url': 'https://www.nytimes.com/2022/03/22/magazine/ethicist-teenage-sex-parenting.html',
  'id': 100000008260402,
  'asset_id': 100000008260402,
  'source': 'New York Times',
  'published_date': '2022-03-22',
  'updated': '2024-04-25 13:35:02',
  'section': 'Magazine',
  'subsection': '',
  'nytdsection': 'magazine',
  'adx_keywords': 'Ethics (Personal);Content Type: Service',
  'column': null,
  'byline': 'By Kwame Anthony Appiah',
  'type': 'Article',
  'title': 'My Daughter Is Having Sex With Her Best Friend. Must I Tell Her Mom?',
  'abstract': 'The magazine\'s Ethicist columnist on whether to keep your teenagerâ€™s.',
  'des_facet': [
    'Ethics (Personal)',
    'Content Type: Service',
  ],
  'org_facet': [],
  'per_facet': [],
  'geo_facet': [],
  'media': [
    {
      'type': 'image',
      'subtype': 'photo',
      'caption': '',
      'copyright': 'Illustration by Tomi Um',
      'approved_for_syndication': 0,
      'media-metadata': [
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-thumbStandard.jpg',
          'format': 'Standard Thumbnail',
          'height': 75,
          'width': 75,
        },
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-mediumThreeByTwo210.jpg',
          'format': 'mediumThreeByTwo210',
          'height': 140,
          'width': 210,
        },
        {
          'url': 'https://static01.nyt.com/images/magazine/27mag-ethicist/27mag-ethicist-mediumThreeByTwo440.jpg',
          'format': 'mediumThreeByTwo440',
          'height': 293,
          'width': 440,
        },
      ],
    },
  ],
  'eta_id': 0,
}]
