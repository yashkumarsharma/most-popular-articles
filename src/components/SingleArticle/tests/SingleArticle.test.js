import React from 'react'
import renderer from 'react-test-renderer'

import SingleArticle from '../SingleArticle'
import { mockArticle } from './mockData'

describe('check SingleArticle', () => {
  it('renders correctly when no article is passed', () => {
    const tree = renderer
      .create(<SingleArticle />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when article is passed', () => {
    const tree = renderer
      .create(<SingleArticle article={mockArticle} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders details correctly', () => {
    const tree = renderer
      .create(<SingleArticle article={mockArticle} />)

    expect(tree.root.findAllByProps({ 'data-testid': 'article-details' }).length).toBe(1)
  })

  it('renders correct published date', () => {
    const tree = renderer
      .create(<SingleArticle article={mockArticle} />)

    expect(tree.root.findByProps({ 'data-testid': 'date-details' }).children)
      .toStrictEqual([ 'Published on ', '22 Mar 2022' ])
  })
})
