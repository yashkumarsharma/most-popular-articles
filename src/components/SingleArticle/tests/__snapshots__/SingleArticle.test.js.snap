// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check SingleArticle renders correctly when article is passed 1`] = `
<div
  className="overflow-y-auto md:w-3/5 bg-gray-100 p-4 md:mt-0 mt-4"
  data-testid="article-details"
>
  <h3
    className="text-2xl font-semibold mb-2"
  >
    My Daughter Is Having Sex With Her Best Friend. Must I Tell Her Mom?
  </h3>
  <div
    className="mb-4 text-center"
  >
    <img
      alt="My Daughter Is Having Sex With Her Best Friend. Must I Tell Her Mom?"
      className="article-image w-full mx-auto rounded-lg shadow-lg"
      src="https://static01.nyt.com/images/2022/03/27/magazine/27mag-ethicist/27mag-ethicist-mediumThreeByTwo440.jpg"
    />
    <div
      className="flex flex-row justify-between max-w-2/3 mx-auto"
    >
      <div
        className="text-sm text-gray-600"
        data-testid="date-details"
      >
        Published on 
        22 Mar 2022
      </div>
      <div
        className="text-sm text-gray-600"
      >
        By Kwame Anthony Appiah
      </div>
    </div>
  </div>
  <p
    className="mb-4"
  >
    The magazine's Ethicist columnist on whether to keep your teenagerâ€™s.
  </p>
  <div
    className="mb-4"
  >
    <h4
      className="font-semibold mb-2"
    >
      Descriptive Terms
    </h4>
    <div
      className="text-sm text-gray-600"
    >
      <div
        className="inline-block bg-gray-200 rounded-full px-3 py-1 m-1"
      >
        <span>
          Ethics (Personal)
        </span>
      </div>
      <div
        className="inline-block bg-gray-200 rounded-full px-3 py-1 m-1"
      >
        <span>
          Content Type: Service
        </span>
      </div>
    </div>
  </div>
  <div
    className="mb-4"
  >
    <h4
      className="font-semibold mb-2"
    >
      Keywords
    </h4>
    <div
      className="text-sm text-gray-600"
    >
      <div
        className="inline-block bg-gray-200 rounded-full px-3 py-1 m-1"
      >
        <span>
          Ethics (Personal)
        </span>
      </div>
      <div
        className="inline-block bg-gray-200 rounded-full px-3 py-1 m-1"
      >
        <span>
          Content Type: Service
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`check SingleArticle renders correctly when no article is passed 1`] = `
<div
  className="overflow-y-auto md:w-3/5 bg-gray-100 p-4 md:mt-0 mt-4"
  data-testid="article-details"
>
  <h3
    className="text-2xl font-semibold mb-2"
  />
  <div
    className="mb-4 text-center"
  >
    <img
      className="article-image w-full mx-auto rounded-lg shadow-lg"
      src=""
    />
    <div
      className="flex flex-row justify-between max-w-2/3 mx-auto"
    >
      <div
        className="text-sm text-gray-600"
        data-testid="date-details"
      >
        Published on 
      </div>
      <div
        className="text-sm text-gray-600"
      />
    </div>
  </div>
  <p
    className="mb-4"
  />
</div>
`;
